Panic in parseElement when element reports extremely large size

I noticed (while accidentally parsing the middle of an MKV container) that it's quite easy to cause a panic in `parseElement` by supplying data that is interpreted as an element with an excessively-large size. The panic occurs in the `data := make([]byte, size)` call in the non-master-element case:

```
panic: runtime error: makeslice: len out of range                                                                                                     
                                                                                                                                                      
goroutine 46 [running]:                                                                                                                               
github.com/remko/go-mkvparse.parseElement(0xe7d5a0, 0xc002816d80, 0x96, 0x0, 0xe964c0, 0xc0003667e0, 0x14, 0x0, 0x0)                                  
        /home/danerat/go/pkg/mod/github.com/remko/go-mkvparse@v0.0.0-20190519181556-e343daff3525/mkvparse.go:133 +0x380                               
github.com/remko/go-mkvparse.parseElements(0xe7d5a0, 0xc002816d80, 0x0, 0xffffffffffffffff, 0x0, 0xe964c0, 0xc0003667e0, 0x4, 0xc002826740, 0x452eef) 
        /home/danerat/go/pkg/mod/github.com/remko/go-mkvparse@v0.0.0-20190519181556-e343daff3525/mkvparse.go:84 +0xce
github.com/remko/go-mkvparse.Parse(0xe7d5a0, 0xc002816d80, 0xe964c0, 0xc0003667e0, 0x13c16a0, 0x79a1c8858d01)
        /home/danerat/go/pkg/mod/github.com/remko/go-mkvparse@v0.0.0-20190519181556-e343daff3525/mkvparse.go:73 +0x6e
...
```

For example, I'm able to trigger this by passing `[]byte{0xa3, 0x01, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}`, which I believe corresponds to a SimpleBlock element of length 2^64-1. This makes it hard to use this package to parse untrusted data.

Would it be reasonable to have a maximum element size? I'm not familiar enough with the MKV spec to know if it would make sense to hardcode a limit in go-mkvparse, or if users should be able to specify one (perhaps via a function on the `Handler` interface, or if you want to avoid non-backward-compatible changes, by making `Parse` accept a new `...Option` arg in the pattern described at e.g. https://www.sohamkamani.com/golang/options-pattern/).

And thanks for making this package! I had trouble finding any other EBML/MKV parsers that would actually let me access raw block data.

--- a/mkvparse.go
+++ b/mkvparse.go
@@ -5,6 +5,7 @@
 package mkvparse
 
 import (
+	"bytes"
 	"encoding/binary"
 	"fmt"
 	"io"
@@ -121,8 +122,9 @@ func parseElement(reader io.Reader, currentOffset int64, level int, handler Hand
 				return -1, err
 			}
 		} else {
-			data := make([]byte, size)
-			reader.Read(data)
+			if _, err := readData(reader, size); err != nil {
+				return -1, err
+			}
 		}
 		err = handler.HandleMasterEnd(id, info)
 		if err != nil {
@@ -130,8 +132,10 @@ func parseElement(reader io.Reader, currentOffset int64, level int, handler Hand
 		}
 		return count, nil
 	} else {
-		data := make([]byte, size)
-		reader.Read(data)
+		data, err := readData(reader, size)
+		if err != nil {
+			return -1, err
+		}
 
 		switch typ {
 		case uintegerType:
@@ -173,6 +177,18 @@ func NameForElementID(id ElementID) string {
 ////////////////////////////////////////////////////////////////////////////////
 
 var baseDate = time.Date(2001, time.January, 1, 0, 0, 0, 0, time.UTC)
+
+// readData reads and returns size bytes from r.
+// An error is returned if EOF is encountered before the requested bytes have been read.
+func readData(r io.Reader, size int64) ([]byte, error) {
+	// Use bytes.Buffer to avoid allocating the full size until needed:
+	// https://github.com/remko/go-mkvparse/issues/4
+	var buf bytes.Buffer
+	if _, err := io.CopyN(&buf, r, size); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
 
 func convertBytesToSignedInt(data []byte) int64 {
 	if data[0] >= 0x80 {

--- a/mkvparse_test.go
+++ b/mkvparse_test.go
@@ -105,6 +105,7 @@ func (p *ParseHandler) HandleBinary(id ElementID, value []byte, info ElementInfo
 type ParseTest struct {
 	data  []byte
 	event ParseEvent
+	fail  bool // error is expected
 }
 
 func TestParseElement(t *testing.T) {
@@ -120,23 +121,37 @@ func TestParseElement(t *testing.T) {
 				},
 				time.Date(1980, time.January, 21, 21, 03, 0, 0, time.UTC),
 			},
+			false,
+		},
+		// Avoid panicking with a too-large slice allocation when an element claims a
+		// very large size: https://github.com/remko/go-mkvparse/issues/4
+		"excessive size": {
+			[]byte{0xa3, 0x01, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x01, 0x02, 0x03, 0x04},
+			ParseEvent{},
+			true,
 		},
 	}
 	for name, test := range tests {
 		reader := bytes.NewReader(test.data)
 		handler := ParseHandler{}
 		count, err := parseElement(reader, 0, 0, &handler)
-		if err != nil {
-			t.Errorf("%s: %v", name, err)
-		}
-		if count != int64(len(test.data)) {
-			t.Errorf("%s: %d != %d. Data: %v", name, count, len(test.data), test.data)
-		}
-		if len(handler.events) != 1 {
-			t.Errorf("%s: Invalid event count: %d", name, len(handler.events))
-		}
-		if test.event != handler.events[0] {
-			t.Errorf("%s: Invalid event: %v != %v", name, test.event, handler.events[0])
+		if test.fail {
+			if err == nil {
+				t.Errorf("%s: Unexpectedly succeeded", name)
+			}
+		} else {
+			if err != nil {
+				t.Errorf("%s: %v", name, err)
+			}
+			if count != int64(len(test.data)) {
+				t.Errorf("%s: %d != %d. Data: %v", name, count, len(test.data), test.data)
+			}
+			if len(handler.events) != 1 {
+				t.Errorf("%s: Invalid event count: %d", name, len(handler.events))
+			}
+			if test.event != handler.events[0] {
+				t.Errorf("%s: Invalid event: %v != %v", name, test.event, handler.events[0])
+			}
 		}
 	}
 }

